### ISSUE TITLE
Broken job arguments after updating to 1.7 version (extra timestamp argument for jobs taking 0 arguments)

### ISSUE THREAD

**Issue by Azdaroth** [ondrejbartas/sidekiq-cron#350]
After updating to 1.7, we've experienced an issue with an extra argument being added for no reason - as far as the config goes, `args` is an empty array, `date_as_argument` is set to false. I suspect it's related to https://github.com/ondrejbartas/sidekiq-cron/pull/329 but haven't managed yet to identify the exact cause behinf it, especially that `enqueue_args` method returns an empty array for affected jobs.

**Comment by leandro** [comment]
I can second this issue as well. Now we're getting the following error when using it:
> ArgumentError
wrong number of arguments (given 1, expected 0)

And taking one of the errors of ours as example, the array below is represents the arguments it received while it wasn't expecting any arguments:
```ruby
[1658897402.8730807]
```

**Comment by nicalpi** [comment]
We have the same issue here, but reverting to 1.6.0 seems to still, somehow, send this argument now. Any one of you experience the same thing?

**Comment by nicalpi** [comment]
Obviously, it was only after having made some code changes that I realised the new `date_as_argument:` in the readme. While this can be set to false, it would have been nice to have it false as a default I think to not break existing implementations.

**Comment by nicalpi** [comment]
@Azdaroth @leandro if like us you upgrade to 1.7.0 before realising the new argument, then I would recommend you to delete all your existing cron jobs and restart sidekiq. This will then recreate the job now setting correctly the `      date_as_argument` as false.

I think, if like us you deployed the code before setting the `date_as_argument:` to  `false` then the jobs were now set in stone with this a lingering argument.

**Comment by julianajuliano** [comment]
We had the same issue and currently have no way of updating to 1.7.0

**Comment by cameronbourgeois** [comment]
It wasn't quite clear from the above comments but from what I can tell, 1.7.0 ignores date_as_argument in your config and sends the additional date parameter to the perform method anyway.

After downgrading to 1.6.0, the `@date_as_argument=false` had disappeared from my jobs, but the argument was then defined explicitly: `@args=[1660264240.1838207]`.

As @nicalpi suggested, destroying and recreating my jobs cleared this arg. This is the command I used:

```
Sidekiq::Cron::Job.destroy_all!
Sidekiq::Cron::Job.load_from_hash YAML.load_file('path/to/my_schedule.yml')
```

Once this was done, I checked my jobs and I see the argument is gone: `@args=[]`.

However, even after restarting sidekiq, I still get jobs failing with the same error.
If I `enque!` the job from the console, the job succeeds, but I am still seeing failures in my logs for the actual scheduled jobs.

Next I will try downgrading back to the version I was at before, which was 1.4.0.

**Comment by markets** [comment]
Hello 👋🏼 and sorry for my late reply, was on holidays! 

I upgraded my apps to v1.7.0 without any problem 🤔, but I was not using this attribute (`date_as_argument`) before the upgrade. I'm also defining the `perform` method with the following signature `def perform(*args)` even for jobs with no arguments, and loading my jobs via the default `config/schedule.yml` autoloader.

Would like to hear @vsuhachev thoughts on this one.

Sorry for any inconvenience 🙏🏼!

**Comment by vsuhachev** [comment]
Hello,

As suggested above, you need to delete and recreate the schedule after switching to version 1.7.0

If there are already created jobs on Sidekiq that have "date_as_argument = true" in the schedule, they should work as expected if the perform method takes this additional parameter. Before 1.7.0, this argument was added when scheduling the cron task. After 1.7.0 it is added when submitting a job to the Sidekiq queue.

If you reproduce the error - give me an example of the schedule file (or the code that creates the schedule). Perhaps you are using ActiveJob jobs or the undocumented "message" parameter when creating the schedule?

**Comment by mensfeld** [comment]
Deleting and recreating did not work for me fyi. For now I had to add `*args` at the end to make it work as expected after the upgrade.

**Comment by vsuhachev** [comment]
> Deleting and recreating did not work for me fyi. For now I had to add `*args` at the end to make it work as expected after the upgrade.

If you reproduce the error - give me an example of the schedule file (or the code that creates the schedule)

**Comment by ychaker** [comment]
@vsuhachev I've been able to reproduce the issue. I deleted the existing cron jobs and reloaded the cron schedule, and I'm still getting the same error.

this is my config file:

```yml
daily_sync_ro_totals:
  cron: "0 1 * * *"
  class: "Daily::SyncRoTotalsJob"
  queue: default

daily_ro_exporter:
  cron: "0 5 * * *"
  class: "Daily::RoExporterJob"
  queue: apis

daily_sync_inventory_with_qbo:
  cron: "30 0 * * * America/Chicago"
  class: "Integrations::QuickbooksOnline::SyncRepairShopsInventoryWorker"
  queue: default
```

and this is the part from my `Gemfile.lock` that might matter:

```
    sidekiq (6.5.6)
      connection_pool (>= 2.2.5)
      rack (~> 2.0)
      redis (>= 4.5.0, < 5)
    sidekiq-batch (0.1.8)
      sidekiq (>= 3)
    sidekiq-client-cli (0.1.7)
      cli
      sidekiq (>= 2.15)
    sidekiq-cron (1.7.0)
      fugit (~> 1)
      sidekiq (>= 4.2.1)
    sidekiq-failures (1.0.4)
      sidekiq (>= 4.0.0)
    sidekiq-unique-jobs (7.1.27)
      brpoplpush-redis_script (> 0.1.1, <= 2.0.0)
      concurrent-ruby (~> 1.0, >= 1.0.5)
      sidekiq (>= 5.0, < 8.0)
      thor (>= 0.20, < 3.0)
```

does this help? 
anything else you need me to provide?

**Comment by 23tux** [comment]
Is there any news on this? We run in the same issue when upgrading to 1.7

**Comment by vsuhachev** [comment]
@ychaker 

1. Has the "date_as_argument" flag been set before upgrade for any job?
2. How did you clear the schedule?
3. What version of Ruby?

It would be helpful to see the sidekiq logs in debug mode.

**Comment by ychaker** [comment]
@vsuhachev 

1. no, we never used that option
2. I used the following, as was suggested by one of the comments above:

```
Sidekiq::Cron::Job.destroy_all!
Sidekiq::Cron::Job.load_from_hash YAML.load_file('path/to/my_schedule.yml')
```

3. Ruby version: `3.1.2`

will be difficult to get you the logs, but I'll try.

should I just add `date_as_argument: false` to the yml file for each of the cron jobs, would that solve the issue, is that the quickest and simplest solution?

**Comment by vsuhachev** [comment]
> should I just add date_as_argument: false to the yml file for each of the cron jobs, would that solve the issue, is that the quickest and simplest solution?

You can try, but there is a check only for "true" values, otherwise it is always false.

What version of redis server and redis gem are you using?

**Comment by ychaker** [comment]
Redis gem: `redis (4.8.0)`
Redis server: `5.0.6`

> You can try, but there is a check only for "true" values, otherwise it is always false.

okay, but doesn't this mean that it's true by default? or is that just an example, and there's no default for that option?

<img width="855" alt="Screen Shot 2022-11-22 at 1 42 47 PM" src="https://user-images.githubusercontent.com/29512/203395721-18f9957a-1041-40ac-ad56-376e84ad326a.png">

**Comment by vsuhachev** [comment]
> okay, but doesn't this mean that it's true by default? or is that just an example, and there's no default for that option?

There are defaults for all options, but here are examples and descriptions of their effects.

I tried to reproduce the bug in your versioned environment and couldn't get it.. 

I had this error only once, when I updated the version and there were tasks in the queue that were scheduled on the old version. After removing these tasks from the queue, the error disappeared.

Also, my schedule is reloaded every time the application is restarted.

Мy code for this:

```ruby
# config/initializers/sidekiq_cron.rb
if Sidekiq.server?
  ["default", Rails.env, "#{Rails.env}.local"]
    .map { |fn| Rails.root.join("config", "schedules", "#{fn}.yml") }
    .select { |fn| File.exist?(fn) }
    .map { |fn| YAML.load_file(fn) }
    .reduce({}, &:merge)
    .then { |schedule| Sidekiq::Cron::Job.load_from_hash!(schedule) }
end
```

**Comment by runephilosof-abtion** [comment]
I also saw this issue.
However, only once on deploying the upgrade. I got some failed queued jobs with the date arguments.
The following enqueued jobs did not have the date argument.
So after deleting the weird jobs with date argument, everything was fine again. :bug:

**Comment by mkusaka** [comment]
Hello.

I also faced this problem and found the possible cause.

Verification environment
- New version of sidekiq-cron: 1.9.1
- Old version of sidekiq-cron: 1.6.0

After some trial and error, we found that this problem is caused by the following behavior

1. newer version of sidekiq-cron registers jobs with date_as_argument option

````
irb(main):002:0> conn = Sidekiq.redis { |conn| conn }
irb(main):003:0> pp conn.hgetall("cron_job:10")
{"date_as_argument"=>"0",
 "description"=>"",
 "active_job"=>"false",
 "status"=>"enabled",
 "queue_name_prefix"=>"",
 "symbolize_args"=>"false",
 "message"=>
  "{\"queue\"\"default\",\"class\":\"YourGreatJob\",\"args\":[]}",
 "last_enqueue_time"=>"2023-01-30 10:40:35 +0000",
 "cron"=>"30 * * * *",
 "queue_name_delimiter"=>"",
 "klass"=>"YourGreatJob",
 "name"=>"10",
 "args"=>"[]"}
````

2. Our application is deployed using a rolling update (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/deployment-type-ecs.html) method. So, There is a temporary time when new and old sidekiq processes are connected to the same redis. In this case, if a sidekiq 1.6.0 sidekiq process executes a job with this new option, it will not be processed properly

It seems that `args["date_as_argument"]` becomes `"0"` and is evaluated as true at the following location at sidekiq-cron 1.6.0

https://github.com/sidekiq-cron/sidekiq-cron/blob/v1.6.0/lib/sidekiq/cron/job.rb#L282

I'll try to submit a modified PR.

**Comment by markets** [comment]
Hello 👋🏼 

Thanks to @mkusaka (#392) we have now a potential fix!

Please, if you can test this branch in your app and report back, that would be really helpful to release a new version:

```ruby
gem 'sidekiq-cron', github: 'mkusaka/sidekiq-cron', branch: 'patch'
```

**Comment by mkusaka** [comment]
FYI: I have already applied the changes to https://github.com/sidekiq-cron/sidekiq-cron/issues/350#issuecomment-1412892389 in my development environment and deployed multiple times, and have confirmed that the errors that used to appear no longer occur!

**Comment by markets** [comment]
Hello @Azdaroth @ychaker @23tux @mensfeld @cameronbourgeois 👋🏼 

Did you had a change to try that patch https://github.com/sidekiq-cron/sidekiq-cron/issues/350#issuecomment-1412892389? It will help a lot to cut a new version 🙏🏼!

**Comment by ychaker** [comment]
@markets sorry, I am unable to try the changes, the company I was working with when I was experiencing this issue got affected by the layoffs, so I no longer have access to that code base or environment.

**Comment by Kayne** [comment]
This is still a ongoing problem.
We are using `sidekiq 7.3.0` and `sidekiq-cron 1.9.3`, we do not use `date_as_argument` and never have been. 
This is sample job from our `schedule.yml`:

```
night_synch:
  class: "NightSynchJob"
  cron: "0 3 * * *"
  queue: backend_low_priority
  active_job: true
```

and this is how job look like when found using `Sidekiq::Cron::Job.find`:

```
 #<Sidekiq::Cron::Job:0x00007f7514d93790
  @active_job=true,
  @active_job_queue_name_delimiter="",
  @active_job_queue_name_prefix="",
  @args=[],
  @cron="0 3 * * *",
  @date_as_argument=false,
  @description="",
  @fetch_missing_args=false,
  @klass="NightSynchJob",
  @last_enqueue_time=nil,
  @message="{\"retry\":false,\"queue\":\"backend_low_priority\",\"class\":\"NightSynchJob\",\"args\":[]}",
  @name="night_synch",
  @queue="backend_low_priority",
  @queue_name_with_prefix="backend_low_priority",
  @status="enabled",
  @symbolize_args=false>
```
The `Sidekiq::Cron::Job` is freshly done, just before the server restart I deleted all of them (`Sidekiq::Cron::Job.destroy_all!`).

As shown above the `date_as_argument` is set to `false` yet still the argument is passed:
<img width="474" alt="Zrzut ekranu 2024-07-23 o 08 10 22" src="https://github.com/user-attachments/assets/0a2a7c7b-6b07-4ea9-a412-bb1a9dd1040e">
